Refs #35438 - Enhance VMware listing commands and switch to cluster_name param

chris1984 opened the pull request:
Requires https://github.com/theforeman/foreman/pull/9383

chris1984 commented:
@ofedoren can this get a review when you get a chance please?

ofedoren commented on lib/hammer_cli_foreman/compute_resource/vmware.rb:
```diff
@@ -64,7 +64,9 @@ def interface_attributes
           Fields::Field.new(:label => _('Datacenter'), :path => [:datacenter]),
-          Fields::Field.new(:label => _('Server'), :path => [:server])
+          Fields::Field.new(:label => _('Server'), :path => [:server]),
+          Fields::Field.new(:label => _('Console password set'), :path => [:set_console_password]),
+          Fields::Field.new(:label => _('Caching enabled'), :path => [:caching_enabled])
```
```suggestion
          Fields::Boolean.new(:label => _('Console password set'), :path => [:set_console_password]),
          Fields::Boolean.new(:label => _('Caching enabled'), :path => [:caching_enabled])
```

ofedoren reviewed (commented):
Thanks, @chris1984! Just a few small nitpicks inline.

Also, I've read initial issue and it seems like in hammer we still have a problem with `cluster_id`:
![ScreenShot-1661786066286](https://user-images.githubusercontent.com/32508194/187237128-feb3f923-d130-4c16-8bc5-71c119d80a6f.png)
![ScreenShot-1661786112506](https://user-images.githubusercontent.com/32508194/187237164-a2e04560-dde5-461b-bd8e-f588185fa621.png)
![ScreenShot-1661786140333](https://user-images.githubusercontent.com/32508194/187237198-065a5e7f-238a-4f86-938b-7edd4573f228.png)
![ScreenShot-1661786161323](https://user-images.githubusercontent.com/32508194/187237234-a49689e3-cef6-4670-91b7-43eb789d32ed.png)

And yes, I was testing this on the latest Foreman's `develop` so your changes are present.

chris1984 commented:
@ofedoren updated.

Changed `cluster_id` to `cluster_name` in the help text, but having that be used as the `cluster_id` since looking into the fog-vsphere code we don't use the raw id for anything. I didn't want to take away the ability to see id's so to prevent confusion and users using the ID instead of the name, I changed the attribute to name but behind the scenes we use that as the ID to they pass in the name.

** Hammer commands with normal cluster structure, with new cluster_name attribute:

*** Networks
```bash
[vagrant@centos7-hammer-devel hammer-cli-foreman]$ hammer compute-resource networks --id 1 --cluster-name "Satellite-Engineering"
----------------|-------------------------------|----------------|------------------|--------
ID              | NAME                          | DATACENTER     | VIRTUAL SWITCH   | VLAN ID
----------------|-------------------------------|----------------|------------------|--------
dvportgroup-572 | ABJ-Network                   | RH_Engineering | Satellite-Switch | 209    
dvportgroup-56  | Host-Network                  | RH_Engineering | Satellite-Switch | 0  
----------------|-------------------------------|----------------|------------------|--------
```

*** Resource pools
```bash
[vagrant@centos7-hammer-devel hammer-cli-foreman]$ hammer compute-resource resource-pools --id 1 --cluster-name "Satellite-Engineering"
-----------|-----------|-----------------------|---------------
ID         | NAME      | CLUSTER               | DATACENTER    
-----------|-----------|-----------------------|---------------
resgroup-8 | Resources | Satellite-Engineering | RH_Engineering
-----------|-----------|-----------------------|---------------
```

*** Storage pods
```bash
[vagrant@centos7-hammer-devel hammer-cli-foreman]$ hammer compute-resource storage-pods --id 1 --cluster-name "Satellite-Engineering"
------------|-------------|---------------
ID          | NAME        | DATACENTER    
------------|-------------|---------------
group-p6123 | NFS-Cluster | RH_Engineering
------------|-------------|---------------
```

*** Storage Domains
```bash
[vagrant@centos7-hammer-devel hammer-cli-foreman]$ hammer compute-resource storage-domains --id 1 --cluster-name "Satellite-Engineering"
----------------|----------------
ID              | NAME           
----------------|----------------
datastore-17566 | ISOS           
datastore-34    | Local-ESXI23   
datastore-3518  | Local-ESXi3
----------------|----------------
```

Hammer commands with nested cluster structure, with new cluster_name attribute:

*** Networks
```bash
[vagrant@centos7-hammer-devel hammer-cli-foreman]$ hammer compute-resource networks --id 2 --cluster-name "Test-Folder/Test-Cluster"
--------------|------------|-------------|----------------|--------
ID            | NAME       | DATACENTER  | VIRTUAL SWITCH | VLAN ID
--------------|------------|-------------|----------------|--------
network-40902 | VM Network | Toledo-Test |                |        
--------------|------------|-------------|----------------|--------
```

*** Resource pools
```bash
[vagrant@centos7-hammer-devel hammer-cli-foreman]$ hammer compute-resource resource-pools --id 2 --cluster-name "Test-Folder/Test-Cluster"
---------------|-----------|--------------------------|------------
ID             | NAME      | CLUSTER                  | DATACENTER 
---------------|-----------|--------------------------|------------
resgroup-40407 | Resources | Test-Folder/Test-Cluster | Toledo-Test
---------------|-----------|--------------------------|------------
```

*** Storage pods
```bash
[vagrant@centos7-hammer-devel hammer-cli-foreman]$ hammer compute-resource storage-pods --id 2 --cluster-name "Test-Folder/Test-Cluster"
---|------|-----------
ID | NAME | DATACENTER
---|------|-----------

```

*** Storage Domains
```bash
[vagrant@centos7-hammer-devel hammer-cli-foreman]$ hammer compute-resource storage-domains --id 2 --cluster-name "Test-Folder/Test-Cluster"
----------------|-----------
ID              | NAME      
----------------|-----------
datastore-40901 | datastore1
----------------|-----------
```

chris1984 commented:
If this looks good to you I will update the tests that failed

ofedoren commented:
Although I'm OK with renaming the option, we would need to deprecate it first instead of removal.

chris1984 commented:
@ofedoren how does this look?

ofedoren commented on lib/hammer_cli_foreman/compute_resource.rb:
```diff
@@ -130,9 +133,24 @@ class AvailableNetworksCommand < HammerCLIForeman::ListCommand
+      option '--cluster-name', "Name of cluster", _("Cluster name to search by"),
+        attribute_name: :option_cluster_id
+
+      option '--cluster-id', "ID of cluster", _("Deprecated, use cluster-name")
```
```suggestion
      option '--cluster-id', 'ID', _(Cluster ID'),
             deprecated: _('Use --cluster-name instead')
```

ofedoren commented on lib/hammer_cli_foreman/compute_resource.rb:
```diff
@@ -215,21 +251,44 @@ class AvailableStorageDomainsCommand < HammerCLIForeman::ListCommand
       action :available_storage_pods
       command_name 'storage-pods'
 
+      option '--cluster-name', "Name of cluster", _("Cluster name to search by"),
+        attribute_name: :option_cluster_id
+
+      option '--cluster-id', "ID of cluster", _("Deprecated, use cluster-name")
```
Those changes are quite repetitive. Let's extract them into a command extension:
```ruby
module HammerCLIForeman
  module CommandExtensions
    class ComputeResourceSubcommand < HammerCLI::CommandExtensions
      option '--cluster-id', 'ID', _('Cluster ID'),
             deprecated: _('Use --cluster-name instead')
      option '--cluster-name', 'NAME', _('Cluster name or path to search by'),
             attribute_name: :option_cluster_id

      request_params do |params|
        params['cluster_id'] = params['cluster_id'].gsub('/', '%2F') if params['cluster_id']
      end
    end
  end
end
```

And then we can update the related commands like:
```ruby
build_options without: :cluster_id

extend_with(
  HammerCLIForeman::CommandExtensions::ComputeResourceSubcommand.new(only: %i[option request_params])
)
```

ofedoren reviewed (changes requested):
Thanks, @chris1984! It works now. Some code-related suggestions which also should fix almost all of the tests.

And can we have at least one more test to cover this specific case?

chris1984 commented:
@ofedoren I updated with your suggestions. When trying to run the test suite locally I get:
```ruby
/home/vagrant/hammer-cli-foreman/lib/hammer_cli_foreman/compute_resource.rb:153:in `<class:AvailableNetworksCommand>': uninitialized constant HammerCLIForeman::CommandExtensions::ComputeResourceSubcommand (NameError)
```

ofedoren commented on lib/hammer_cli_foreman/compute_resource.rb:
```diff
@@ -133,9 +136,19 @@ class AvailableNetworksCommand < HammerCLIForeman::ListCommand
       end
 
-      build_options
+      def request_params
+        params = super
+        params['cluster_id'] = params['cluster_id'].gsub('/', '%2F')
+        params
+      end
```
This is no longer needed since it's present in the extension.

ofedoren commented on lib/hammer_cli_foreman/compute_resource.rb:
```diff
@@ -206,9 +224,18 @@ class AvailableResourcePoolsCommand < HammerCLIForeman::ListCommand
       end
 
-      build_options
+      def request_params
+        params = super
+        params['cluster_id'] = params['cluster_id'].gsub('/', '%2F')
+        params
+      end
```
Same as above

ofedoren commented on lib/hammer_cli_foreman/compute_resource.rb:
```diff
@@ -220,7 +247,14 @@ class AvailableStorageDomainsCommand < HammerCLIForeman::ListCommand
       end
 
-      build_options
+      def request_params
+        params = super
+        params['cluster_id'] = params['cluster_id'].gsub('/', '%2F')
+        params
+      end
```
Same as above

ofedoren commented on lib/hammer_cli_foreman/compute_resource.rb:
```diff
@@ -230,9 +264,17 @@ class AvailableStoragePodsCommand < HammerCLIForeman::ListCommand
       end
 
-      build_options
+      def request_params
+        params = super
+        params['cluster_id'] = params['cluster_id'].gsub('/', '%2F')
+        params
+      end
```
Same as above

ofedoren reviewed (commented):
Thanks, @chris1984, you need to also register the extension in https://github.com/theforeman/hammer-cli-foreman/blob/master/lib/hammer_cli_foreman/command_extensions.rb

chris1984 commented:
@ofedoren how does this look? I kept it in a separate commit so you can review it easier. If it all looks good I can squash it or I think you can squash them all too.

ofedoren reviewed (approved):
Thanks, @chris1984, here is :cookie: for you :)
